Dev analysis

jfcabana opened the pull request:
Several impovements and functionnality additions to the analysis module. The code still lacks proper documentation, but the addition made are useful, thus the pull request.

**Update analysis.py**
- Added an 'apply' (default=True) argument to apply_factor_from_roi method to allow the use of this method to only get and print the scale factor in an ROI, and not applying this factor to the film.
- Changed get_profile_offsets method so that the profile position is taken from the show_results window instead of being at image center.
- Added show_results_ontype method.
In the show_results() window, the get_profile_offsets workflow is launched, at the current x/y position, when 'enter' is pressed.
- Added 'plot_isodoses' method in DoseAnalysis class to show a 2D plot of isodoses (film dose and ref dose on same graph).
- Added a tool to be able to draw an interactive ruler in the graphs. Used in the plot_isodoses method in analysis for now. This tool is defined in a new "tools" module.
- Improved analysis report, including a contours plot showing 2D isodoses of film dose and reference dose.
- Added analyse_clusters method, which identifies the clusters of dose on the reference dose map, based on a given threshold relative to maximum dose. For each cluster found, interactive plots are shown to successively get median dose over an roi inside the cluster, and x/y profiles offset and width difference. Theses values are stored in the object under self.clusters_analysis. When publishing pdf report, if the argument plot_clusters_analysis=True, a page is added with details for each cluster. Goal would be to automate this analysis, but not ready for now.

**Update imageRGB.py**
- Baseclass method plotCB() is replaced by an argument to the plot() method instead, as plotCB() was just a duplicate of plot() with the addition of a line to add the colorbar. plotCB() is left there for now for backward compatibility, until we change to v2.
- Added 'plot_isodoses' method in BaseImage class to perform a contours plot.
- Added a clusters identification in the BaseImage class (detect_clusters method). The method returns a list of center of masses identified in the image, based on a threshold value (applied to the maximum value in the image)

jfcabana committed:
Update analysis.py

Added an 'apply' (default=True) argument to apply_factor_from_roi method to allow the use of this method to only get and print the scale factor in an ROI, and not applying this factor to the film.

jfcabana committed:
Merge branch 'heads/OMG_master' into dev_analysis

jfcabana committed:
Update analysis.py

Changed get_profile_offsets method so that the profile position is taken from the show_results window instead of being at image center.

jfcabana committed:
Create load_saved_analysis.py

Script pour charger fichier analyse.pkl avec un filedialog

jfcabana committed:
Update analysis.py

Added show_results_ontype method.
In the show_results() window, the get_profile_offsets workflow is launched, at the current x/y position, when 'enter' is pressed.

jfcabana committed:
Deprecated plotCB

in module imageRGB, the Basecalss method plotCB() is replaced by an argument to the plot() method instead, as plotCB() was just a duplicate of plot() with the addition of a line to add the colorbar.

plotCB() is left there for now for backward compatibility, until we change to v2.

jfcabana committed:
Isodose plot

Added 'plot_isodoses' method in imageRGB BaseImage class to perform a contours plot.

Added 'plot_isodoses' method in analysis DoseAnalysis class to show a 2D plot of isodoses (film dose and ref dose on same graph).

jfcabana committed:
Ruler

Added a tool to be able to draw an interactive ruler in the graphs. Used in the plot_isodoses method in analysis for now.

This tool is defined in a new "tools" module.

jfcabana committed:
Improved analysis report

Some improvements made to allow easier analysis of single isocenter multi-meta dose distribution.

Added a clusters identification in the imageRGB module BaseImage class (detect_clusters method). The method returns a list of center of masses identified in the image, based on a threshold value (applied to the maximum value in the image)

Added some stuff to the PDF report:
- A new page with the 2D isodoses plot
- If profiles_coord='detect' argument passed to the publish_pdf method, dose clusters are identified on the reference dose distribution. A new page is then added to the report for each dose cluster identified showing the horizontal an vertical profiles at this position.

jfcabana committed:
Clusters analysis

Added analyse_clusters method, which identifies the clusters of dose on the reference dose map, based on a given threshold relative to maximum dose.

For each cluster found, interactive plots are shown to successively get median dose over an roi inside the cluster, and x/y profiles offset and width difference.

Theses values are stored in the object under self.clusters_analysis.

When publishing pdf report, if the argument plot_clusters_analysis=True, a page is added with details for each cluster.

Goal would be to automate this analysis, but not ready for now.

jfcabana committed:
Pimped report and cleaning

jfcabana committed:
Update analysis.py

jfcabana committed:
v1.7.0

jfcabana committed:
Update analysis.py

Fixed indentation

jfcabana committed:
Update analysis.py

petruong commented on src/omg_dosimetry/analysis.py:
```diff
@@ -619,31 +636,33 @@ def plot_profile(self, ax=None, profile='x', position=None, title=None, diff=Fal
+            film_prof, ref_prof = film[:,position], ref[:,position]
         
         x_axis = (np.array(range(0, len(film_prof))) / self.film_dose.dpmm).tolist()
-        y_max = max(np.concatenate((film_prof, ref_prof)))
```
In my opinion, y_max calculation remains relevant in the case that the film dose is higher than the reference dose. 
This is also a dynamic y_max calculation as it is based on the position of the crosshair (x and y), which allows the profile y max to change as the position of the crosshair is selected (as opposed to ax.set_ylim((0,self.ref_dose.array.max() * 1.05)) in code line 682 which causes the profile to remain at a static max y).

jfcabana replied:
I can't recall why I did this. I think I needed to generate different profile plots while keeping the same overall ylim. I will add the option to do both.

petruong replied:
Pay careful attention that I have re-organized the order of the xlim and ylim definitions (lines: **674-677**) in order for the vertical line call (**679-681**) to take in the updated xlim and ylim definition.

petruong commented on src/omg_dosimetry/analysis.py:
```diff
@@ -690,27 +737,34 @@ def show_results(self, fig=None, x=None, y=None, show = True):
             at a given x/y coordinates, and draw lines on the dose distribution maps
             to show where the profile is taken.
         """
-        self.plot_profile(ax=axes[-2], profile='x', position=[x, y])
```
Did the function call for plot_profile really have to change (parameter call for position variable [x, y] as opposed to just x or y) as show_profiles function call already requires both x and y variables as it takes that information from the crosshair position coordinates (set_profile function).

jfcabana replied:
Yes, for the PDF reports.

petruong commented on src/omg_dosimetry/analysis.py:
```diff
@@ -653,8 +672,34 @@ def plot_profile(self, ax=None, profile='x', position=None, title=None, diff=Fal
             ax_diff.plot(x_axis, diff_prof,'g-', linewidth=0.25)
+            
+        if vertical_line:
+            ax.plot((vertical_line / self.film_dose.dpmm, vertical_line / self.film_dose.dpmm), ax.get_ylim(), 'k:', linewidth = 1)
```
Would be fixed (dynamic range) with: 
if vertical_line: ax.plot((vertical_line / self.film_dose.dpmm, vertical_line / self.film_dose.dpmm), 
                                  (0, y_max * 1.05), 'k:', linewidth = 1)

jfcabana replied:
ax.get_ylim() makes sure the vertical line spans the whole y range of the plot, whatever the limits are.

petruong commented on src/omg_dosimetry/analysis.py:
```diff
@@ -653,8 +672,34 @@ def plot_profile(self, ax=None, profile='x', position=None, title=None, diff=Fal
+        if vertical_line:
+            ax.plot((vertical_line / self.film_dose.dpmm, vertical_line / self.film_dose.dpmm), ax.get_ylim(), 'k:', linewidth = 1)
+            
+        if xlim:
```
Unsure of the reasoning for the inclusion of xlim and ylim changes.

jfcabana replied:
I needed those for the improved PDF reports. When "clusters analysis" is used, I generate new pages showing the analysis on individual clusters. I plot isodoses and profile plots that are focused on the cluster (instead of showing profiles that span the whole film).

jfcabana replied:
![image](https://github.com/jfcabana/omg_dosimetry/assets/136498453/73b3a6e2-8965-4054-9eea-bef264b2f32b)

petruong commented on src/omg_dosimetry/analysis.py:
```diff
@@ -619,31 +636,33 @@ def plot_profile(self, ax=None, profile='x', position=None, title=None, diff=Fal
+            vertical_line : int, optional
+                If set to True, a dashed vertical line is plotted on the profile at this position
+                
+            xlim : tuple, optional
```
Missing ylim commentary/description (if needed).

jfcabana replied:
Indeed. Commit coming up!

LuisOlivaresJ commented:
Looks great! This weekend I going to read the modifications on imageRGB.py

LuisOlivaresJ committed:
documentation: lut size is (6, nDoses, npixels) and (6, nDoses) with and without lateral correction.

LuisOlivaresJ committed:
Merge pull request #49 from jfcabana/lut_size

documentation: lut size

LuisOlivaresJ commented:
> Looks great! This weekend I going to read the modifications on imageRGB.py

plot_isodoses works on the demo file "A1A_Multi_6cm_001.tif" :smiley:

![Contour_example](https://github.com/jfcabana/omg_dosimetry/assets/41806420/7561b62a-2243-4c8d-94f5-704241001518)

This is the code to reproduce:

```python
from omg_dosimetry.tiff2dose import Gaf
from omg_dosimetry import tiff2dose
import matplotlib.pyplot as plt

# Load demo LUT
path_to_lut_file = tiff2dose.from_demo_lut()

# Load folder to demo tif file
path_to_tif_folder = tiff2dose.from_demo_image()

# Gaf init.
fit_type = 'rational'
clip = 500
gaf = Gaf(
    path_to_tif_folder,
    lut_file=path_to_lut_file,
    fit_type=fit_type,
    clip=clip,
    img_filt=3,
    crop_edges=0.1,
)

# Plotting
fig, axes = plt.subplots(1, 2)
gaf.dose_opt.plot(ax=axes[0], show=False, colorbar=True, title="Dose")
gaf.dose_opt.plot_isodoses(levels=[350], inline=True, ax=axes[1], show=False, title="Contour")

plt.show()
```

LuisOlivaresJ committed:
documentation: detect_clusters

LuisOlivaresJ commented:
Could it be useful to have a method to remove clusters with an area below a given value? Using [this](https://scikit-image.org/docs/stable/api/skimage.measure.html#skimage.measure.regionprops) can be a solution. I am working on it.

jfcabana commented:
> Could it be useful to have a method to remove clusters with an area below a given value? Using [this](https://scikit-image.org/docs/stable/api/skimage.measure.html#skimage.measure.regionprops) can be a solution. I am working on it.

Yes I think it's a good idea!

LuisOlivaresJ committed:
Merge pull request #50 from jfcabana/clusters

detect_clusters docstring

LuisOlivaresJ committed:
enhancement: Added discard_small_clusters() method in BaseImage class to remove small clusters

petruong commented on scripts/analysis/load_saved_analysis.py:
Just to note: I believe that this file was loosely based off of the /tools/Analysis_Pickle.py. 
Originally, it was used for developmental/testing purposes for the .pkl file generated after performing analysis (hence the name Analysis_Pickle); however, this can in theory be used for any other .pkl file generated (i.e., calibration or tiff2dose) and then used further on for debugging purposes. 
So, not limited to just Analysis .pkl files.

jfcabana committed:
Update analysis.py

Added new 'auto' option for ylim in plot_profile(), previous one renamed as 'max'.

jfcabana committed:
Update analysis.py

jfcabana committed:
Merge pull request #51 from jfcabana/clusters

New discard_small_clusters()

petruong committed:
analysis.py Cleanup

Quick minor changes to analysis.py:
- Version date update
- Removed "from scipy import ndimage" as ndimage was not used, and is already referenced with line 21: import scipy.ndimage.filters as spf
- Removed commented out lines of old code (i.e., lines 160) as this code has already been versioned in past commits and can be referenced from those.
- Re-organized the xlim and ylim declaration to be before the vertical line call (line 679) as the parameter ax.get_ylim() needs to occur after the ylim has been updated.
- Cleaned up empty space.

OMG_CISSSO_test:
- Included a revision of the test script used at CISSSO as a reference of how the script is implemented on our side (give some perspective of the tools/methodology used in analysis.py).

petruong reviewed (approved):
Note: I have made a few changes (review my recent commit to analysis.py): mostly formatting changes. 
Tested and it's functional on my side, but should verify on your side that nothing breaks. 

Clusters I have not tested as I haven't dived too deep on this at our center. 
Perhaps @LuisOlivaresJ could redo the test/example case with the updated "remove small clusters" implementation.

LuisOlivaresJ commented:
> Note: I have made a few changes (review my recent commit to analysis.py): mostly formatting changes. Tested and it's functional on my side, but should verify on your side that nothing breaks.
> 
> Clusters I have not tested as I haven't dived too deep on this at our center. Perhaps @LuisOlivaresJ could redo the test/example case with the updated "remove small clusters" implementation.

"remove_small_clusters" runs without errors :thumbsup:. Thanks @petruong
